containers-tests/tests/IntMapValidity.hs

```diff
@@ -12,54 +13,59 @@ import Utils.Containers.Internal.BitUtil (bitcount)
 valid :: IntMap a -> Property
 valid t =
   counterexample "nilNeverChildOfBin" (nilNeverChildOfBin t) .&&.
-  counterexample "commonPrefix" (commonPrefix t) .&&.
-  counterexample "maskRespected" (maskRespected t)
+  counterexample "prefixOk" (prefixOk t)
 
 -- Invariant: Nil is never found as a child of Bin.
 nilNeverChildOfBin :: IntMap a  -> Bool
 nilNeverChildOfBin t =
   case t of
     Nil -> True
     Tip _ _ -> True
-    Bin _ _ l r -> noNilInSet l && noNilInSet r
+    Bin _ l r -> noNilInSet l && noNilInSet r
   where
     noNilInSet t' =
       case t' of
         Nil -> False
         Tip _ _ -> True
-        Bin _ _ l' r' -> noNilInSet l' && noNilInSet r'
+        Bin _ l' r' -> noNilInSet l' && noNilInSet r'
 
--- Invariant: The Mask is a power of 2. It is the largest bit position at which
---            two keys of the map differ.
-maskPowerOfTwo :: IntMap a -> Bool
-maskPowerOfTwo t =
+-- Invariants:
+-- * All keys in a Bin start with the Bin's Prefix.
+-- * All keys in the Bin's left child have the Prefix's mask bit unset.
+-- * All keys in the Bin's right child have the Prefix's mask bit set.
+prefixOk :: IntMap a -> Property
+prefixOk t =
   case t of
-    Nil -> True
-    Tip _ _ -> True
-    Bin _ m l r ->
-      bitcount 0 (fromIntegral m) == 1 && maskPowerOfTwo l && maskPowerOfTwo r
+    Nil -> property ()
+    Tip _ _ -> property ()
+    Bin p l r ->
+      let px = unPrefix p
+          m = px .&. (-px)
+          keysl = keys l
+          keysr = keys r
+          debugStr = concat
+            [ "px=" ++ showIntHex px
```

Comment:
Yes, but binary is just too verbose. For example, we might get a failure like
```
    *** Failed! Falsified (after 8 tests and 9 shrinks):
    -1
    fromList [(0,()),(1,())]
    prefixOk
    px=0b1, keysl=[0b0], keysr=[0b1111111111111111111111111111111111111111111111111111111111111111,0b1]
    prefix not shared
```
With hex, that is
```
    px=0x1, keysl=[0x0], keysr=[0xffffffffffffffff,0x1]
```

Reply:
fits in the terminal width with binary no? I'm not sure which one would be better but i sure like that byte chunks would be separated.
```
0b11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111
or
0xff ff ff ff ff ff ff ff
```

Reply:
Ugh, yeah, you're right. In an ideal world, it would be great to have a debug build that uses `Int8` and shows things in binary, but that sounds like more trouble than it's worth for the moment.